Add episode 05 (Functions). (#6)
mkcor/python-prog: _episodes/05-functions.md

---
title: "Writing Functions"
teaching: 10
exercises: 15
questions:
- "How can I create my own functions?"
objectives:
- "Explain and identify the difference between function definition and function call."
- "Write a function that takes a small, fixed number of arguments and produces a single result."
keypoints:
- "Break programs down into functions to make them easier to understand."
- "Define a function using `def` with a name, parameters, and a block of code."
- "Defining a function does not run it."
- "Arguments in a function call are matched to its defined parameters."
- "Functions may return a result to their caller using `return`."
---
## Break programs down into functions to make them easier to understand.

*   Human beings can only keep a few items in working memory at a time.

> *   Understand larger/more complicated ideas by understanding and combining pieces.

Reviewer: ```suggestion
*   Handle larger/more complicated ideas by understanding and combining smaller pieces.
```

    print('Hello!')
    print('The weather is nice today.')
    print('Right?')
~~~
{: .language-python}

## Defining a function does not run it.

*   Defining a function does not run it.
    *   Like assigning a value to a variable.
*   Must call the function to execute the code it contains.

~~~
print_greeting()
~~~
{: .language-python}
~~~
Hello!
The weather is nice today.
Right?
~~~
{: .output}

## Arguments in a function call are matched to its defined parameters.

*   Functions are most useful when they can operate on different data.
*   Specify *parameters* when defining a function.
    *   These become variables when the function is executed.
    *   Are assigned the arguments in the call (i.e., the values passed to the function).
    *   If you don't name the arguments when using them in the call, the arguments will be matched to parameters in the order the parameters are defined in the function.

~~~
def print_date(year, month, day):
    joined = str(year) + '/' + str(month) + '/' + str(day)
    print(joined)

print_date(1871, 3, 19)
~~~
{: .language-python}
~~~
1871/3/19
~~~
{: .output}

Or, we can name the arguments when we call the function, which allows us to

> specify them in any order and adds clarity to the call site; otherwise as

Reviewer: ```suggestion
specify them in any order and adds clarity to the call site; otherwise, as
```
PS: What's the 'call *site*,' please?

Author: you are right, not fully clear. I updated it to call statement, does it make more sense?

> one is reading the code they might forget if the second argument is the month

Reviewer: ```suggestion
one is reading the code, they might forget if the second argument is the month
```

> or the day for example.

Reviewer: ```suggestion
or the day, for example.
```

~~~
print_date(month=3, day=19, year=1871)
~~~
{: .language-python}
~~~
1871/3/19
~~~
{: .output}

*   Via [Twitter](https://twitter.com/minisciencegirl/status/693486088963272705):
    `()` contains the ingredients for the function
    while the body contains the recipe.

## Functions may return a result to their caller using `return`.

*   Use `return ...` to give a value back to the caller.
*   May occur anywhere in the function.
*   But functions are easier to understand if `return` occurs:
    *   At the start to handle special cases.
    *   At the very end, with a final result.

~~~
def average(values):
    if len(values) == 0:
        return None
    return sum(values) / len(values)
~~~
{: .language-python}

~~~
a = average([1, 3, 4])
print('average of actual values:', a)
~~~
{: .language-python}
~~~
average of actual values: 2.6666666666666665
~~~
{: .output}

~~~

> print('average of empty list:', average([]))

Reviewer: ```suggestion
b = average([])
print('average of empty list:', b)
```
Just to parallel the previous code snippet. Alternatively, change the latter to `print('average of actual values:', average([1, 3, 4]))`.

~~~
{: .language-python}
~~~
average of empty list: None
~~~
{: .output}

> *   Remember: [every function returns something]({{ page.root }}/04-built-in/).

Reviewer: Oops! We don't have a `04-built-in/` folder in the repo... Please update.

Author: good catch

~~~
{: .language-python}
~~~
1871/3/19
result of call is: None
~~~
{: .output}

> ## Identifying Syntax Errors
>
> 1. Read the code below and try to identify what the errors are
>    *without* running it.
> 2. Run the code and read the error message.
>    Is it a `SyntaxError` or an `IndentationError`?
> 3. Fix the error.
> 4. Repeat steps 2 and 3 until you have fixed all the errors.
>
> ~~~
> def another_function
>   print("Syntax errors are annoying.")
>    print("But at least python tells us about them!")
>   print("So they are usually not too hard to fix.")
> ~~~
> {: .language-python}
>
> > ## Solution
> >
> > ~~~
> > def another_function():
> >   print("Syntax errors are annoying.")
> >   print("But at least Python tells us about them!")
> >   print("So they are usually not too hard to fix.")
> > ~~~
> > {: .language-python}
> {: .solution}
{: .challenge}

> ## Definition and Use
>
> What does the following program print?
>
> ~~~
> def report(pressure):
>     print('pressure is', pressure)
>
> print('calling', report, 22.5)
> ~~~
> {: .language-python}
> > ## Solution
> >
> > ~~~
> > calling <function report at 0x7fd128ff1bf8> 22.5
> > ~~~ 
> > {: .output}
> >

> > > A function call always needs parenthesis, otherwise you get memory address of the function object. So, if we wanted to call the function named report, and give it the value 22.5 to report on, we could have our function call as follows

Reviewer: ```suggestion
> > A function call always needs parentheses, otherwise we get the memory address of the function object. So, if we wanted to call the function named `report`, and give it the value 22.5 to report on, we could have our function call as follows:
```

> > ~~~
> > print("calling")
> > report(22.5)
> > ~~~
> > {: .language-python}
> > ~~~
> > calling
> > pressure is 22.5
> > ~~~
> > {: .output}
> {: .solution}
{: .challenge}

> ## Order of Operations
>
> 1. What's wrong in this example?
>
>     ~~~
>     result = print_time(11, 37, 59)
>
>     def print_time(hour, minute, second):
>        time_string = str(hour) + ':' + str(minute) + ':' + str(second)
>        print(time_string)
>     ~~~
>     {: .language-python}
> 
> 2. After fixing the problem above, explain why running this example code:
>
>     ~~~
>     result = print_time(11, 37, 59)
>     print('result of call is:', result)
>     ~~~
>     {: .language-python}
>
>     gives this output:
>
>     ~~~
>     11:37:59
>     result of call is: None
>     ~~~
>     {: .output}
>
> 3. Why is the result of the call `None`?
>
> > ## Solution
> > 
> > 1. The problem with the example is that the function `print_time()` is defined *after* the call to the function is made. Python

> > > doesn't know how to resolve the name `print_time` since it hasn't been defined yet and will raise a `NameError` e.g.,

Reviewer: ```suggestion
> > doesn't know how to resolve the name `print_time` since it hasn't been defined yet and will raise a `NameError`, i.e.,
```

> > of the `result` variable.
> >
> > 3. `print_time()` does not explicitly `return` a value, so it automatically returns `None`.
> >
> {: .solution}
{: .challenge}

> ## Encapsulation
>
> Fill in the blanks to create a function that takes a list of numbers as an argument,
> and returns the square of the maximum number in that list.
>
> ~~~
>
> def square_of_max(____):
>     data_max = ____
>     return ____
> ~~~
> {: .language-python}
> > ## Solution
> >
> > ~~~
> > 
> > def square_of_max(number_list):
> >     data_max = max(number_list)
> >     return data_max**2
> > ~~~
> > {: .language-python}
> {: .solution}
{: .challenge}

> ## Find the First
>
> Fill in the blanks to create a function that takes a list of numbers as an argument
> and returns the first negative value in the list.
> What does your function do if the list is empty? What if the list has no negative numbers?
>
> ~~~
> def first_negative(values):
>     for v in ____:
>         if ____:
>             return ____
> ~~~
> {: .language-python}
> > ## Solution
> >
> > ~~~
> > def first_negative(values):
> >     for v in values:
> >         if v < 0:
> >             return v
> > ~~~
> > {: .language-python}

> > > If an empty list or a list with all positive values is passed to this function, it returns `None`:

Reviewer: ```suggestion
> > If an empty list or a list with all non-negative values is passed to this function, it returns `None`:
```

>
> > ## Solution
> >
> > ~~~
> > def get_egg_label(mass):
> >     # egg sizing machinery prints a label
> >     egg_label = "Unlabelled"
> >     if mass >= 90:
> >         egg_label = "warning: egg might be dirty"
> >     elif mass >= 85:
> >         egg_label = "jumbo"
> >     elif mass >= 70:
> >         egg_label = "large"
> >     elif mass < 70 and mass >= 55:
> >         egg_label = "medium"
> >     elif mass < 50:
> >         egg_label = "too light, probably spoiled"
> >     else:
> >         egg_label = "small"
> >     return egg_label
> > ~~~
> > {: .language-python}
> {: .solution}
{: .challenge}

> ## Simulating a dynamical system
>
> In mathematics, a [dynamical system](https://en.wikipedia.org/wiki/Dynamical_system) is a system
> in which a function describes the time dependence of a point in a geometrical space. A canonical
> example of a dynamical system is the [logistic map](https://en.wikipedia.org/wiki/Logistic_map),

> > a growth model that computes a new population density (between  0 and 1) based on the current

Reviewer: ```suggestion
> a growth model that computes a new population density (between 0 and 1) based on the current
```

> density. In the model, time takes discrete values 0, 1, 2, ...
>
> 1. Define a function called `logistic_map` that takes two inputs: `x`, representing the current
>    population (at time `t`), and a parameter `r = 1`. This function should return a value 
>    representing the state of the system (population) at time `t + 1`, using the mapping function:
>
>    `f(t+1) = r * f(t) * [1 - f(t)]`
>
> 2. Using a `for` or `while` loop, iterate the `logistic_map` function defined in part 1, starting
>    from an initial population of 0.5, for a period of time `t_final = 10`. Store the intermediate
>    results in a list so that after the loop terminates you have accumulated a sequence of values

> >    representing the state of the logistic map at times `t = [0,1,...,t_final]` (11 values in total).

Reviewer: ```suggestion
>    representing the state of the logistic map at times `t = [0, 1, ..., t_final]` (11 values in total).
```

>    Print this list to see the evolution of the population.
>
> 3. Encapsulate the logic of your loop into a function called `iterate` that takes the initial

> >    population as its first input, the parameter `t_final` as its second input and the parameter

Reviewer: Is it obvious that 'input' can be used as a synonym for 'argument?' Since input is more generic, I would tend to think of 'input' as the set of all arguments (so the first argument would be the 'first piece of input' and so on).

I think it's fine to say: "Here, we are using 'input' as a synonym for 'argument.'"

Author: I think it's rather clear from the uses of word input in this tutorial, but I will update it as you suggest

Reviewer: Now I'm thinking that the word 'parameter' may cause confusion:

* There are 'parameters' in a function's definition in the context of this lesson (in computer programming);
* but, in this sentence, it seems that 'parameter' is used as in *mathematics*, referring to the `r` parameter in this equation `f(t+1) = r * f(t) * [1 - f(t)]`, for example.

For clarity, I think I would simplify the sentence to something like "... write a function ... where the second parameter is `t_final` ..."

>    `r` as its third input. The function should return the list of values representing the state of

> >    the logistic map at times `t = [0,1,...,t_final]`. Run this function for periods `t_final = 100`

Reviewer: ```suggestion
>    the logistic map at times `t = [0, 1, ..., t_final]`. Run this function for periods `t_final = 100`
```

> a growth model that computes a new population density (between 0 and 1) based on the current
> density. In the model, time takes discrete values 0, 1, 2, ...
>
> 1. Define a function called `logistic_map` that takes two inputs: `x`, representing the current
>    population (at time `t`), and a parameter `r = 1`. This function should return a value 
>    representing the state of the system (population) at time `t + 1`, using the mapping function:
>
>    `f(t+1) = r * f(t) * [1 - f(t)]`
>
> 2. Using a `for` or `while` loop, iterate the `logistic_map` function defined in part 1, starting
>    from an initial population of 0.5, for a period of time `t_final = 10`. Store the intermediate
>    results in a list so that after the loop terminates you have accumulated a sequence of values
>    representing the state of the logistic map at times `t = [0, 1, ..., t_final]` (11 values in total).
>    Print this list to see the evolution of the population.
>
> 3. Encapsulate the logic of your loop into a function called `iterate` that takes the initial
>    population as its first input, the parameter `t_final` as its second input and the parameter

> >    `r` as its third input. Note that here, we are using 'input' as a synonym for 'argument The function should return the list of values representing the state of

Reviewer: ```suggestion
>    `r` as its third input. Note that here, we are using 'input' as a synonym for 'argument.' The function should return the list of values representing the state of
```
